### PR TITLE
Make the code more sync

### PR DESCRIPTION
**Pull request by warpech**
Alternative solution to PR #9

### CODE REVIEW

**Review by miyconst — CHANGES_REQUESTED** [review]
@warpech unfortunately I can't accept this pull request.

The `<juicy-select>` was explicitly developed to keep the `<select>` element in the light DOM. 

After your rework the following construction will no longer work:

```html
<juicy-select>
    <select class="custom-class" on-change="onSelectChangeCustom"></select>
</juicy-select>
```

**Comment by warpech** [comment]
Thanks, now I understand the motivation.

This is quite against the common practice in custom elements. Normally, we would do it like this then:

```html
<juicy-select-bind-options options="{{options}}">
    <select slot="select" class="custom-class" on-change="onSelectChangeCustom"></select>
</juicy-select-bind--ptions>
```

I will make a change.

**Comment by miyconst** [comment]
>This is quite against the common practice in custom elements. Normally, we would do it like this then:

Could be. This is how the `pikaday` decorator works, and it's very nice. The element may be renamed, but renaming itself won't solve the issue.

Note, that `<juicy-select>` allows a `<select>` in it's light DOM, but does not require it. If no `<select>` element provided, one will be created.

**Comment by warpech** [comment]
Could you please re-review?

---

> Could be. This is how the `pikaday` decorator works, and it's very nice. The element may be renamed, but renaming itself won't solve the issue.

Yes, I mean this as a potential new custom element. But we don't have resources to work on it now.

File: juicy-select.html [hunk 1]
```diff
@@ -27,45 +27,41 @@
...
+                }
+
+                this.generateOptions();
+
+                if (this.value != this.$select.value && !this.multiple) {
+                    this.set("value", this.$select.value);
                 }
             },
             generateOptions: function () {
+                if (!this.options || !this.options.length) {
```
Comment (miyconst) [inline juicy-select.html]: In this case the empty option or caption text will not be shown when there are no options available, but it should be.

**Review by miyconst — CHANGES_REQUESTED** [review]

File: juicy-select.html (hunk 1 above)
Comment (warpech) [reply juicy-select.html]: Fixed. I thought this is not intended use case, but I agree I should not make such breaking changes.

**Review by miyconst — APPROVED** [review]

**Comment by eriksunsol** [comment]
As an alternative solution to PR #9, this does not work. See https://github.com/Juicy/juicy-select/pull/9#issuecomment-341411932.

File: juicy-select.html [hunk 2]
```diff
@@ -78,8 +83,12 @@
 
                     this.$select.appendChild(opt);
                 }
+                console.log(this.$select.childNodes.length)
```
Comment (warpech) [inline juicy-select.html]: Please remove `console.log`

File: juicy-select.html [hunk 3]
```diff
@@ -38,19 +38,24 @@
...
+                    this.$select.addEventListener('change', this.selectChange);
                 }
-                this.set("multiple", this.$select.hasAttribute("multiple"));
+                else {
+                    this.set("multiple", this.$select.hasAttribute("multiple"));
+                }
+                if (this.options.length && this.$select) {
+                    this.generateOptions();
+                }
+                this.addEventListener("change", this.selectChange);
```
Comment (warpech) [inline juicy-select.html]: Do you need both event listeners, one on line 41 and one on line 49?

File: test/helpers.js [hunk 4]
```diff
@@ -0,0 +1,7 @@
+function simulateChangeEvent(element) {
+  if ('createEvent' in document) {
+    var evt = document.createEvent('HTMLEvents');
+    evt.initEvent('change', false, true);
```
Comment (warpech) [inline test/helpers.js]: `initEvent` is [deprecated](https://developer.mozilla.org/en-US/docs/Web/API/Event/initEvent).

Pls use something like:

```js
var evt = new Event('change', {bubbles: true, composed: true});
element.dispatchEvent(evt);
```

File: test/simple.html [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+                domBind.set('model', model);
+            });
+            it("Should contain correct number of <option> elements: options.length + 1 (for caption)", function () {
+                expect(juicySelect.$select.childNodes.length).to.equal(juicySelect.options.length + 1);
+            });
+            it("Each <option>.innerHTML should equal model.options[index][text-property]", function () {
+                const childNodes = Object.entries(juicySelect.$select.childNodes);
+
+                for (let [index, option] of childNodes) {
+                    index = ~~index;
```
Comment (warpech) [inline test/simple.html]: Please don't use bitwise sorcery. 🧙 Do you [mean Math.floor](https://stackoverflow.com/questions/5971645/what-is-the-double-tilde-operator-in-javascript)?

Comment (alshakero) [reply test/simple.html]: It's a fast alternative to parseInt. But you're right.

File: juicy-select.html (hunk 3 above)
Comment (alshakero) [reply juicy-select.html]: Good catch.

**Comment by alshakero** [comment]
I'll merge and release a minor.

**Comment by alshakero** [comment]
Released here and on NPM.

**Comment by warpech** [comment]
Setting as “Finished” because it is:
- released (with release notes) https://github.com/Juicy/juicy-select/releases/tag/1.1.0

**Pull request merged**